primitives/npos-elections/solution-type/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+/// struct TestSolution {
+///     voters1: vec![(V, T), ..]
+///     voters2: vec![V, P, [(T1, P1), (T2, P2)], T_last]
+///     voters3: <stripped>
+///     voters4: <stripped>
+/// }
+///
+/// impl SolutionBase for TestSolution {};
+/// impl Solution for TestSolution {};
+/// ```
+///
+/// The given struct provides function to convert from/to `Assignment` as part of
+/// [`sp_npos_elections::Solution`] trait:
+///
+/// - `fn from_assignment<..>(..)`
+/// - `fn into_assignment<..>(..)`
+///
+/// ## Compact Encoding
+///
+/// The generated struct is by default deriving both `Encode` and `Decode`. This is okay but could
+/// lead to many `0`s in the solution. If prefixed with `#[compact]`, then a custom compact encoding
+/// for numbers will be used, similar to how `parity-scale-codec`'s `Compact` works.
+///
+/// ```
+/// # use sp_npos_elections_solution_type::generate_solution_type;
+/// # use sp_npos_elections::Solution;
+/// # use sp_arithmetic::per_things::Perbill;
+/// generate_solution_type!(
+///     #[compact]
+///     pub struct TestSolutionCompact::<VoterIndex = u16, TargetIndex = u8, Accuracy = Perbill>(8)
+/// );
+/// ```
+///
+/// ## Pagination
+///
+/// All of the above examples generate one struct that implements [`sp_npos_elections::Solution`]
+/// and [`sp_npos_elections::SolutionBase`]. These solutions do not have pagination. To support
+/// pagination, the `#[pages(_)]` attribute can be used. For example:
+///
+/// ```
+/// # use sp_npos_elections_solution_type::generate_solution_type;
+/// # use sp_npos_elections::Solution;
+/// # use sp_arithmetic::per_things::Perbill;
+/// generate_solution_type!(
+///     #[compact]
+///     #[pages(3)]
+///     pub struct MultiPageSolution::<VoterIndex = u16, TargetIndex = u8, Accuracy = Perbill>(8)
+/// );
+/// ```
+///
+/// Will generate two structs:
+///
+/// 1. `MultiPageSolutionPage` which implements [`sp_npos_elections::Solution`] and
+///    [`sp_npos_elections::SolutionBase`].
+/// 2. `MultiPageSolution` which implements [`sp_npos_elections::MultiPageSolution`].
+///
+/// and `MultiPageSolution` wraps 3 pages of `MultiPageSolutionPage`, in other words:
+///
+/// ```ignore
+/// struct MultiPageSolutionPage { ... }
+/// struct MultiPageSolution {
+/// 	page0: MultiPageSolutionPage { .. },
+/// 	page1: MultiPageSolutionPage { .. },
+/// 	page2: MultiPageSolutionPage { .. },
+/// }
+/// ````
+/// The page count must always be greater than 2.
+#[proc_macro]
+pub fn generate_solution_type(item: TokenStream) -> TokenStream {
+	let SolutionDef {
+		vis,
+		ident,
+		count,
+		voter_pages,
+		voter_type,
+		target_type,
+		weight_type,
+		compact_encoding,
+	} = syn::parse_macro_input!(item as SolutionDef);
+
+	let imports = imports().unwrap_or_else(|e| e.to_compile_error());
+
+	let def = match voter_pages {
+		None => single_page::generate(
+			vis.clone(),
+			ident.clone(),
+			count,
+			voter_type.clone(),
+			target_type.clone(),
+			weight_type.clone(),
+			compact_encoding,
+		)
```

Comment:
Not that important, but what do you think of using as `struct GenerateOpts` to pass these arguments? In JS, normally over 5 arguments I use an object to pass arguments in so you can't make a mistake just by misordering arguments

Reply:
excellent suggestion.